examples/dpo_demo_gemma3.ipynb

```diff
@@ -0,0 +1,849 @@
+    "\n",
+    "from tunix.examples.data import translation_dataset as data_lib\n",
+    "from tunix.generate import sampler as sampler_lib\n",
+    "from tunix.models.gemma3 import params_safetensors as params_safetensors_lib\n",
+    "\n",
+    "from datasets import load_dataset\n",
+    "from tunix.sft.dpo.dpo_trainer import DpoTrainingConfig\n",
+    "from tunix.sft.dpo.dpo_trainer import DpoTrainer\n",
+    "from tunix.sft.dpo.dpo_trainer import TrainingInput\n",
+    "from huggingface_hub import snapshot_download\n",
+    "from tunix.sft.dpo.dpo_trainer import _generate_ids_and_masks\n",
+    "from tunix.models.gemma3 import model as gemma3_model_lib\n",
+    "from datasets import concatenate_datasets\n",
+    "import numpy as np"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# ====== Data ======\n",
+    "TRAIN_DATA_DIR = \"./data/train\"\n",
+    "TEST_DATA_DIR = \"./data/test\"\n",
+    "TRAIN_FRACTION = 1.0\n",
+    "\n",
+    "INTERMEDIATE_CKPT_DIR = \"/content/intermediate_ckpt/\"\n",
+    "# ====== LoRA ======\n",
+    "RANK = 8\n",
+    "ALPHA = 16.0\n",
+    "\n",
+    "# ====== Sharding ======\n",
+    "MESH = [(1, 1), (\"fsdp\", \"tp\")]\n",
+    "\n",
+    "MAX_PROMPT_LENGTH = 192\n",
+    "TOTAL_GENERATION_STEPS = 192\n",
+    "TEMPERATURE = 0.7\n",
+    "TOP_P = 1.0\n",
+    "TOP_K = 50\n",
+    "BETA = 0.1\n",
+    "\n",
+    "# === AdamW, warmup, cosine scheduler ===\n",
+    "LEARNING_RATE = 3e-6\n",
+    "B1 = 0.9\n",
+    "B2 = 0.99\n",
+    "WEIGHT_DECAY = 0.1\n",
+    "\n",
+    "# == Cosine decay with warmup scheduler ==\n",
+    "# Linearly increase learning rate from 0. to 5e-6 in the first 10% training\n",
+    "# steps, and then gradually decrease the learning rate to 0 using cosine\n",
+    "# scheduler.\n",
+    "BATCH_SIZE = 1\n",
+    "NUM_BATCHES = 512\n",
+    "NUM_TEST_BATCHES = 100\n",
+    "EVAL_EVERY_N_STEPS = 100\n",
+    "\n",
+    "NUM_EPOCHS = 1  # can potentially train for more epochs\n",
+    "TRAIN_FRACTION = 1.0\n",
+    "MAX_STEPS = int(NUM_BATCHES * TRAIN_FRACTION * NUM_EPOCHS)\n",
+    "\n",
+    "WARMUP_STEPS = 0.1 * MAX_STEPS\n",
+    "# == Grad clipping ==\n",
+    "# Grad clipping to prevent large gradients. Found this\n",
+    "# important to keep KL divergence in check.\n",
+    "MAX_GRAD_NORM = 0.1\n",
+    "\n",
+    "# ====== Inference ======\n",
+    "GENERATION_CONFIGS = {\n",
+    "    # greedy search\n",
+    "    \"greedy\": {\"temperature\": 1e-4, \"top_k\": 1, \"top_p\": 1.0},\n",
+    "    # some randomness\n",
+    "    \"standard\": {\"temperature\": 0.7, \"top_k\": 50, \"top_p\": 0.95},\n",
+    "    # liberal\n",
+    "    \"liberal\": {\"temperature\": 0.85, \"top_k\": 2000, \"top_p\": 1.0},\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "hnxxLtLAv66G"
+   },
+   "source": [
+    "# Load reference model and LoRA model"
```

Comment:
An explanation here would be nice...can talk about the reference model staying fixed during training, and the LoRA model is the policy model, which is trained, reference model being a Gemma 3 1B model, etc.

Can talk about the r